ethcore/src/private_transactions/mod.rs

```diff
@@ -0,0 +1,819 @@
+		let private = PrivateTransaction {
+			encrypted: encrypted_transaction,
+			contract: *contract,
+		};
+		let private_state = self.execute_private_transaction(BlockId::Latest, &transaction)?;
+		trace!("Private transaction created, encrypted transaction: {:?}, private state: {:?}", private, private_state);
+		let contract_validators = self.get_validators(BlockId::Latest, &contract)?;
+		trace!("Required validators: {:?}", contract_validators);
+		self.transactions_for_signing.lock().add_transaction(private.hash(), transaction, contract_validators, private_state)?;
+		self.broadcast_private_transaction(private.rlp_bytes().into_vec());
+		Ok(())
+	}
+
+	/// Extract signed transaction from private transaction
+	fn extract_original_transaction(&self, private: PrivateTransaction, contract: &Address) -> Result<UnverifiedTransaction, EthcoreError> {
+		let encrypted_transaction = private.encrypted.clone();
+		let transaction_bytes = self.decrypt(contract, &encrypted_transaction)?;
+		let original_transaction: UnverifiedTransaction = UntrustedRlp::new(&transaction_bytes).as_val()?;
+		Ok(original_transaction)
+	}
+
+	/// Process received private transaction
+	pub fn import_private_transaction(&self, rlp: &[u8]) -> Result<(), EthcoreError> {
+		let validator_accounts = self.config.lock().validator_accounts.clone();
+		trace!("Private transaction received");
+		if validator_accounts.is_empty() {
+			self.broadcast_private_transaction(rlp.into());
+			return Ok(());
+		}
+		let private_tx: PrivateTransaction = UntrustedRlp::new(rlp).as_val()?;
+		let contract = private_tx.contract;
+		let contract_validators = self.get_validators(BlockId::Latest, &contract)?;
+		let addresses_intersection = contract_validators
+			.iter()
+			.filter(|&&address| validator_accounts
+				.iter()
+				.find(|validator| validator.address == address)
+				.is_some())
+			.collect::<Vec<_>>();
+		if addresses_intersection.is_empty() {
+			// Not for verification, broadcast further to peers
+			self.broadcast_private_transaction(rlp.into());
+			return Ok(());
+		}
+		else {
+			trace!("Private transaction taken for verification");
+			let client = self.client.read().clone().ok_or_else(|| PrivateTransactionError::ClientIsMalformed)?;
+			let client = client.upgrade().ok_or_else(|| PrivateTransactionError::ClientIsMalformed)?;
+			let original_tx = self.extract_original_transaction(private_tx.clone(), &contract)?;
+			let details_provider = TransactionDetailsProvider::new(&*client as &MiningBlockChainClient);
+			let insertion_time = client.chain_info().best_block_number;
+			// Verify with the first account available
+			let validation_account = addresses_intersection[0].clone();
+			trace!("The following account will be used for verification: {:?}", validation_account);
+			let add_res = self.transactions_for_verification.lock().add_transaction(original_tx, contract, validation_account, private_tx.hash(), &details_provider, insertion_time);
+			match add_res {
```

Comment:
More "rusty" way to write this would be `add_res.and_then(|_| <your Ok branch here>)`
This will keep the `Err` branch without explicit rewriting.

Please also note that `add_res` can be trivially inlined after those changes.

Reply:
This was done intentionally in order to release mutex.